Use docvalue skippers on dimension fields

romseygeek opened the pull request:
Disable points and terms indexes on dimension fields, and use doc_values 
skippers instead.  Dimension fields are part of the `_tsid` and so align
with the index sorts, meaning that doc_values skippers will work well for
querying and filtering, and have a much smaller disk footprint than full 
indexes.

romseygeek committed:
Default indexed to false for keyword dimension fields

Commit:
[CI] Auto commit changes from spotless

felixbarny committed:
Merge branch 'main' into dimensions/sparse-indexes

felixbarny commented on server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:
```diff
@@ -279,21 +279,20 @@ private Builder(
 
             this.dimension = TimeSeriesParams.dimensionParam(m -> toType(m).fieldType().isDimension()).addValidator(v -> {
-                if (v && (indexed.getValue() == false || hasDocValues.getValue() == false)) {
+                if (v && (hasDocValues.getValue() == false)) {
```
Can we at least also add `IpFieldMapper`? I think that will be important for the experimentation.

romseygeek committed:
Same for IpFieldMapper

romseygeek committed:
Merge remote-tracking branch 'romseygeek/dimensions/sparse-indexes' into dimensions/sparse-indexes

Commit:
[CI] Auto commit changes from spotless

felixbarny commented:
I've extracted flame graphs for two queries that have regressed the most:

[avg_avgot_cpu_load_1m_filtered_by_hosts_5m-main.html](https://github.com/user-attachments/files/23105754/avg_avgot_cpu_load_1m_filtered_by_hosts_5m-main.html)
[avg_avgot_cpu_load_1m_filtered_by_hosts_5m-skippers.html](https://github.com/user-attachments/files/23105758/avg_avgot_cpu_load_1m_filtered_by_hosts_5m-skippers.html)

[avg_avgot_cpu_load_1m_prefix_by_hosts_5m-main.html](https://github.com/user-attachments/files/23105759/avg_avgot_cpu_load_1m_prefix_by_hosts_5m-main.html)
[avg_avgot_cpu_load_1m_prefix_by_hosts_5m-skippers.html](https://github.com/user-attachments/files/23105764/avg_avgot_cpu_load_1m_prefix_by_hosts_5m-skippers.html)


What both have in common that they're using a two phase iterator in the skippers version. The `prefix_by_hosts` query is also doing a lot of UTF-8 to UTF-16 conversions that it didn't do before.

felixbarny committed:
Merge branch 'main' into dimensions/sparse-indexes

romseygeek committed:
Merge remote-tracking branch 'origin/main' into dimensions/sparse-indexes

romseygeek committed:
Merge remote-tracking branch 'origin/main' into dimensions/sparse-indexes

romseygeek committed:
wip

romseygeek committed:
Use IndexType in TermBasedFieldType constructors

In preparation for doc-value skippers being enabled in some keyword
field types, TermBasedFieldType should take IndexType directly in
its constructor rather than booleans for indexed and docvalues.

romseygeek committed:
Merge branch 'kft/constructor-index-type' into dimensions/sparse-indexes

romseygeek committed:
Use sparse indexes for dimension fields

romseygeek committed:
Merge remote-tracking branch 'origin/main' into dimensions/sparse-indexes

romseygeek committed:
Merge remote-tracking branch 'origin/main' into dimensions/sparse-indexes

romseygeek committed:
iter

romseygeek committed:
Merge remote-tracking branch 'romseygeek/dimensions/sparse-indexes' into dimensions/sparse-indexes

romseygeek committed:
Merge branch 'main' into dimensions/sparse-indexes

romseygeek commented:
I just realised that I haven't included `flattened` as part of this PR.  It does look more complicated, so maybe should be left for a separate follow-up.

romseygeek committed:
precommit

felixbarny reviewed (approved):
LGTM!
+1 to handle flattened fields separately (if at all)

romseygeek committed:
Merge remote-tracking branch 'romseygeek/dimensions/sparse-indexes' into dimensions/sparse-indexes

romseygeek committed:
iter

romseygeek committed:
iter

romseygeek committed:
Merge remote-tracking branch 'origin/main' into dimensions/sparse-indexes

elasticsearchmachine commented:
Pinging @elastic/es-storage-engine (Team:StorageEngine)

romseygeek committed:
Merge branch 'main' into dimensions/sparse-indexes

martijnvg reviewed (approved):
LGTM 👍

romseygeek committed:
Merge branch 'main' into dimensions/sparse-indexes
